File: parkinglot.client/src/app/models/vehicle.ts
import { ParkingSection } from "./parking-section";
import { SpotSize } from "./spot-size";
import { VehicleType } from "./vehicle-type";

export interface Vehicle {
  id?: number;
  plate: string;
  
  typeId: number;
  sectionId?: number;

  type?: VehicleType;
  section?: ParkingSection;
}

export function sectionSpotsUsed(section?: ParkingSection, types?: VehicleType[], sizes?: SpotSize[]): number {
  return !section || !types || !sizes ? 0 : section?.vehicles?.reduce((spots, vehicle) => {
    return spots += sizes?.find(s => s.id === types?.find(t => t.id === vehicle.typeId)?.id)?.vehicleUsage?.find(s => s.spotSizeId === section?.sizeId)?.usage ?? 0;
  }, 0) ?? 0;
}

File: ParkingLot.Server/Models/Vehicle.cs
namespace ParkingLot.Server.Models
{
    public class Vehicle : Entity
    {
        public string Plate { get; set; }

        public int TypeId { get; set; }
        public int? SectionId { get; set; }

        public VehicleType? Type { get; set; }
        public ParkingSection? Section { get; set; }
    }
}
